pgctl/service.py

```diff
@@ -215,17 +215,53 @@ def background(self):
             return
 
         with self.flock() as lock:
-            log = self.path.join('log').open('a')
-            log = prepend_timestamps_to(log)
+            log_fifo_path = self.path.join('log_pipe').strpath
+            if not os.path.exists(log_fifo_path):
+                os.mkfifo(log_fifo_path)
+
+            # The writer (opened as RW to avoid blocking) must be opened
+            # before checking if the logging process is running,
+            # since we're guaranteed* that if a writing pipe is open,
+            # and the logger is running, it will remain running.
+            #
+            # Thus, we can maintain the assertion that if we find the logger
+            # is running, it will remain running.
+            #
+            # * assuming the process isn't manually killed
+            log_fifo_writer = os.open(log_fifo_path, os.O_RDWR)
+            if not self.is_logger_running():
+
+                with open(os.devnull, 'w') as devnull:
+                    log_fifo_reader = os.open(log_fifo_path, os.O_RDONLY)
+                    Popen(
+                        (
+                            's6-log',
+                            'n5',
+                            's10485760',  # 10MB
+                            'T',
+                            self.path.join('logs').strpath,
+                        ),
+                        stdin=log_fifo_reader,
+                        stdout=devnull.fileno(),
+                        stderr=devnull.fileno(),
+
+                        # This dies automatically when stdin closes, so we
+                        # don't need to maintain a lock
+                        close_fds=True,
```

Comment:
Per offline discussion, we have to drop the lock. This is because we need to be able to restart the service while the logger is running, which happens when poll-ready execs `pgctl restart` (so it can't hold the lock otherwise this deadlocks).